File: src/FootballManagerApp/FootballManagerApp.Shared/Responses/ApiResponse.cs
using System.Text.Json.Serialization;

namespace FootballManagerApp.Shared.Responses;

public record ApiResponse<T>
{
    public int Status { get; init; }
    public string Message { get; init; } = string.Empty;
    public T? Data { get; init; }

    [JsonPropertyName("_links")]
    public Dictionary<string, HateoasLink> Links { get; init; } = new();

    public static ApiResponse<T> Success(
        T data, string message = "OK", int status = 200) =>
        new() { Status = status, Message = message, Data = data };

    public static ApiResponse<T> Created(
        T data, string message = "Creado correctamente") =>
        new() { Status = 201, Message = message, Data = data };

    public static ApiResponse<T> NotFound(
        string message = "No encontrado") =>
        new() { Status = 404, Message = message, Data = default };

    public static ApiResponse<T> BadRequest(
        string message = "Solicitud inválida") =>
        new() { Status = 400, Message = message, Data = default };

    public static ApiResponse<T> Unauthorized(
        string message = "No autorizado") =>
        new() { Status = 401, Message = message, Data = default };

    public static ApiResponse<T> Forbidden(
        string message = "Sin permisos") =>
        new() { Status = 403, Message = message, Data = default };

    public static ApiResponse<T> ServerError(
        string message = "Error interno del servidor") =>
        new() { Status = 500, Message = message, Data = default };

    public static ApiResponse<T> NotImplemented(
        string message = "Funcionalidad no disponible") =>
        new() { Status = 501, Message = message, Data = default };

    public static ApiResponse<T> Conflict(
        string message = "Conflicto") =>
        new() { Status = 409, Message = message, Data = default };

    public static ApiResponse<T> NoContent() =>
        new() { Status = 204, Message = "Sin contenido", Data = default };

    public ApiResponse<T> WithLinks(Dictionary<string, HateoasLink> links) =>
        this with { Links = links };
}

File: backend-node/src/utils/apiResponse.ts
/**
 * Espejo bit-a-bit del ApiResponse<T> / PagedResponse<T> del backend .NET.
 * Cualquier endpoint debe responder con esta forma para que el toggle del
 * frontend sea transparente.
 */

export type Links = Record<string, { href: string; rel: string; method: string }>;

export interface ApiResponse<T> {
  status: number;
  message: string;
  data: T | null;
  _links: Links;
}

export interface PagedResponse<T> {
  status: number;
  message: string;
  data: T[];
  page: number;
  limit: number;
  total: number;
  pages: number;
  _links: Links;
}

// ─────────────── 2xx ───────────────

export const ok = <T>(
  data: T,
  message = 'OK',
  links: Links = {},
): ApiResponse<T> => ({ status: 200, message, data, _links: links });

export const created = <T>(
  data: T,
  message = 'Creado correctamente',
  links: Links = {},
): ApiResponse<T> => ({ status: 201, message, data, _links: links });

export const noContent = (): ApiResponse<null> => ({
  status: 204, message: 'Sin contenido', data: null, _links: {},
});

export const multiStatus = <T>(
  data: T,
  message: string,
  links: Links = {},
): ApiResponse<T> => ({ status: 207, message, data, _links: links });

// ─────────────── 4xx ───────────────

export const badRequest = (message = 'Solicitud inválida'): ApiResponse<null> => ({
  status: 400, message, data: null, _links: {},
});

export const unauthorized = (
  message = 'No autorizado',
): ApiResponse<null> => ({ status: 401, message, data: null, _links: {} });

export const forbidden = (
  message = 'Sin permisos',
): ApiResponse<null> => ({ status: 403, message, data: null, _links: {} });

export const notFound = (
  message = 'No encontrado',
): ApiResponse<null> => ({ status: 404, message, data: null, _links: {} });

export const conflict = (
  message = 'Conflicto',
): ApiResponse<null> => ({ status: 409, message, data: null, _links: {} });

export const unprocessable = (
  message = 'Entidad no procesable',
): ApiResponse<null> => ({ status: 422, message, data: null, _links: {} });

// ─────────────── 5xx ───────────────

export const serverError = (
  message = 'Error interno',
): ApiResponse<null> => ({ status: 500, message, data: null, _links: {} });

export const badGateway = (
  message = 'Error al conectar con servicio externo',
): ApiResponse<null> => ({ status: 502, message, data: null, _links: {} });

export const serviceUnavailable = (
  message = 'Servicio temporalmente no disponible',
): ApiResponse<null> => ({ status: 503, message, data: null, _links: {} });

export const gatewayTimeout = (
  message = 'Timeout',
): ApiResponse<null> => ({ status: 504, message, data: null, _links: {} });

// ─────────────── Paginado ───────────────

export const paged = <T>(
  data: T[],
  page: number,
  limit: number,
  total: number,
  message = 'OK',
  links: Links = {},
): PagedResponse<T> => ({
  status: 200,
  message,
  data,
  page,
  limit,
  total,
  pages: limit <= 0 ? 0 : Math.ceil(total / limit),
  _links: links,
});
